### DISCUSSION TITLE
Blank screen after OTA update

### DISCUSSION THREAD

**Discussion by barteld** [serkri/SmartEVSE-3#217] [Q&A] [answered]
I tried doing an over the air update but somehow that failed and I was greeted with a blank screen. After that I installed the serial drivers for windows, and was able to do a usb update (after setting the framework to  https://github.com/espressif/arduino-esp32#idf-release/v4.4 as building from the master failed). However the blank screen remains and connection to the serial port I see an endless reboot loop:
rst:0x10 (RTCWDT_RTC_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:1184
load:0x40078000,len:13260
load:0x40080400,len:3028
entry 0x400805e4
ets Jul 29 2019 12:21:46

rst:0x10 (RTCWDT_RTC_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:1184
load:0x40078000,len:13260
load:0x40080400,len:3028
entry 0x400805e4
ets Jul 29 2019 12:21:46

Any ideas as to what I'm doing wrong?

**Comment by dingo35** [comment]
I would go back to the original firmware first, remember you have to flash both firmware.bin AND spiffs.bin  . Use prebuilt version to be sure you do not upload a wrongly compiled version.

Also this thread might have some pointers for you,:
https://github.com/SmartEVSE/SmartEVSE-3/issues/21

**Comment by dingo35** [comment]
Also: try to surf to http://ipaddress/update, you might get lucky and get the OTA flash screen......

**Comment by dingo35** [comment]
Now if you're in a reboot cycle that last remark probably wont help you.

I updated platformio.ini, if you wish to compile yourself just run:
pio run -t upload 

...from the directory where platformio.ini is located. You might need to edit the upload-device line in that file....

**Comment by JohanEH22** [comment]
I got the same problem! Also with the mstegen-version.. guess it has to do with a broken/changed dependency? 

HowTo load the prebuild (usb)
copy firmware.bin/spiffs.bin from latest release to SmartEVSE-3\.pio\build\release, and then:
pio run -t nobuild -t upload


About the not working bin-file: my build looks like this, @dingo35 could you share yours?:

Processing release (platform: espressif32; board: esp32dev; framework: arduino)
----------------------------------------------------------------------------------------------------------------------------------------------------Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/esp32dev.html
PLATFORM: Espressif 32 (6.4.0) > Espressif ESP32 Dev Module
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
DEBUG: Current (cmsis-dap) External (cmsis-dap, esp-bridge, esp-prog, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, tumpa)
PACKAGES:
 - framework-arduinoespressif32 @ 2.0.14+sha.44da992
 - tool-esptoolpy @ 1.40501.0 (4.5.1)
 - toolchain-xtensa-esp32 @ 8.4.0+2021r2-patch5
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 44 compatible libraries
Scanning dependencies...
Dependency Graph
|-- ESPAsyncDNSServer @ 1.0.1+sha.3ccb2d3
|-- ESPAsyncWebServer-esphome @ 3.1.0
|-- ESPAsync_WiFiManager @ 1.14.1
|-- eModbus @ 1.7.0
|-- ArduinoJson @ 6.21.3
|-- MQTT @ 2.5.1+sha.7afcfb1
|-- Preferences @ 2.0.0
|-- AsyncTCP-esphome @ 2.0.1
|-- ESP Async WebServer @ 1.2.3
|-- ESPmDNS @ 2.0.0
|-- FS @ 2.0.0
|-- SPI @ 2.0.0
|-- SPIFFS @ 2.0.0
|-- Update @ 2.0.0
|-- WiFi @ 2.0.0
|-- RemoteDebug @ 3.0.5
Building in release mode
Retrieving maximum program size .pio\build\release\firmware.elf
Checking size .pio\build\release\firmware.elf
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [==        ]  15.9% (used 52172 bytes from 327680 bytes)
Flash: [=======   ]  66.4% (used 1174625 bytes from 1769472 bytes)

**Comment by dingo35** [comment]
Processing release (platform: espressif32; board: esp32dev; framework: arduino)
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/esp32dev.html
PLATFORM: Espressif 32 (6.2.0+sha.efc0303) > Espressif ESP32 Dev Module
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
DEBUG: Current (cmsis-dap) External (cmsis-dap, esp-bridge, esp-prog, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, tumpa)
PACKAGES: 
 - framework-arduinoespressif32 @ 2.0.14+sha.44da992 
 - tool-esptoolpy @ 1.40501.0 (4.5.1) 
 - toolchain-xtensa-esp32 @ 8.4.0+2021r2-patch5
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 44 compatible libraries
Scanning dependencies...
Dependency Graph
|-- ESPAsyncDNSServer @ 1.0.1+sha.3ccb2d3
|-- ESPAsyncWebServer-esphome @ 3.1.0
|-- ESPAsync_WiFiManager @ 1.14.1
|-- eModbus @ 1.7.0
|-- ArduinoJson @ 6.21.3
|-- MQTT @ 2.5.1+sha.7afcfb1
|-- Preferences @ 2.0.0
|-- AsyncTCP-esphome @ 2.0.1
|-- ESP Async WebServer @ 1.2.3
|-- ESPmDNS @ 2.0.0
|-- FS @ 2.0.0
|-- SPI @ 2.0.0
|-- SPIFFS @ 2.0.0
|-- Update @ 2.0.0
|-- WiFi @ 2.0.0
|-- RemoteDebug @ 3.0.5
Building in release mode
Retrieving maximum program size .pio/build/release/firmware.elf
Checking size .pio/build/release/firmware.elf
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [==        ]  15.9% (used 52124 bytes from 327680 bytes)
Flash: [=======   ]  66.4% (used 1174325 bytes from 1769472 bytes)

**Comment by JohanEH22** [comment, marked as answer]
diff johan.txt dingo.txt

johan: PLATFORM: Espressif 32 (6.4.0) > Espressif ESP32 Dev Module

dingo: PLATFORM: Espressif 32 (6.2.0+sha.efc0303) > Espressif ESP32 Dev Module


_platformio.ini_

[env:release]
**platform = espressif32**
platform_packages =
   framework-arduinoespressif32 @ https://github.com/espressif/arduino-esp32#2.0.14
extra_scripts = pre:eModbus-fix/fix.py
build_flags = 
	-DLOG_LEVEL=5
        -Wall
        -Wextra
        -Wunused-variable

**Reply by barteld** [reply]
Thanks! that did the trick. I was messing about with low level esptool commands but this is much easier.

**Reply by JohanEH22** [reply]
That's great! 
You are referring to:
pio run -t nobuild -t upload

Or did you get the compiled version working?

**Reply by barteld** [reply]
I got the compiled version working (platformio run --target upload). That is to say: I see a "ready  to charge" screen again. Haven't actually tried charging yet. Will do in an hour or so. 
Edit: it's charging. Note this is not with the fork but with the original mstegen one. Will try that again later on.

**Comment by dingo35** [comment]
Do I understand correctly that I could prevent this problem if I change:
platform=espressif32
to
platform=espressif32 @ v6.2.0

(and perhaps follow docs to upgrade to v6.4.0) ?

**Reply by barteld** [reply]
I'm quite new to this so I frankly don't know exactly what's the difference between the platform and the platform-packages. All I can say is that it started working again when I changed the platform-packages line to this:

platform = espressif32
platform_packages =  framework-arduinoespressif32 @ https://github.com/espressif/arduino-esp32#2.0.14

I suspect changing the platform will also work.

**Reply by dingo35** [reply]
That change was already made a few days ago in the repository.....

**Comment by dingo35** [comment]
Do I understand correctly that I could prevent this problem if I change:
platform=espressif32
to
platform=espressif32 @ v6.2.0

(and perhaps follow docs to upgrade to v6.4.0) ?

**Comment by negronoir** [comment]
I don't know if it has anything to do with it, but the problem that I have detected is that I'm compiling on a Windows OS. When compiling everything works perfectly but the image, once uploaded to the SmartEVSE, does not work and leaves the screen blank.

I see that the original firmware.bin image has a size of 1,169,680 bytes
However, when compiling exactly the same source code on Windows, the image has a size of 1,180,816 bytes

I've read about "stripping symbols" procedure shall be done on Windows OS platformio images to get the same size and alignment of Linux ones and make then work right, but I'm not sure about the process

This is my build in Windows:

Processing release (platform: espressif32; board: esp32dev; framework: arduino)
------------------------------------------------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/esp32dev.html
PLATFORM: Espressif 32 (6.4.0) > Espressif ESP32 Dev Module
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
DEBUG: Current (cmsis-dap) External (cmsis-dap, esp-bridge, esp-prog, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, tumpa)
PACKAGES:
 - framework-arduinoespressif32 @ 2.0.14+sha.44da992
 - tool-esptoolpy @ 1.40501.0 (4.5.1)
 - toolchain-xtensa-esp32 @ 8.4.0+2021r2-patch5
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 44 compatible libraries
Scanning dependencies...
Dependency Graph
|-- ESPAsyncDNSServer @ 1.0.1+sha.3ccb2d3
|-- ESPAsyncWebServer-esphome @ 3.1.0
|-- ESPAsync_WiFiManager @ 1.14.1
|-- eModbus @ 1.7.0
|-- ArduinoJson @ 6.21.3
|-- MQTT @ 2.5.1+sha.7afcfb1
|-- Preferences @ 2.0.0
|-- AsyncTCP-esphome @ 2.0.1
|-- ESP Async WebServer @ 1.2.3
|-- ESPmDNS @ 2.0.0
|-- FS @ 2.0.0
|-- SPI @ 2.0.0
|-- SPIFFS @ 2.0.0
|-- Update @ 2.0.0
|-- WiFi @ 2.0.0
|-- RemoteDebug @ 3.0.5
Building in release mode
Retrieving maximum program size .pio\build\release\firmware.elf
Checking size .pio\build\release\firmware.elf
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [==        ]  15.9% (used 52172 bytes from 327680 bytes)
Flash: [=======   ]  66.4% (used 1174513 bytes from 1769472 bytes)
============================================= [SUCCESS] Took 10.13 seconds =============================================

**Reply by negronoir** [reply]
I answer myself: That's not the problem.
I just compiled the image in CentOS and it has a different size than the one in Windows but when I upload it, the screen also gets blank.

I cannot find the difference between distributed release img which works and image from code compiled at home neither Windows nor Linux which doesn't work

**Comment by JohanEH22** [comment]
Well, tried all sort of versions but no success. Spiffs works, change in index.html, upload and go. But firmware.bin no go. When using usb the unit hangs, when using the upload it seems not to load (i changed in evse.cpp the text at line 3696 to  
doc["version"] = "johan" + String(VERSION);
so /settings should give something with johan. No luck however. When uploading the stegen binary it immediate runs..

I think it has something to do with the platformio environment or so:
@dingo35 what is your version?
$ pio --version
PlatformIO Core, version 6.1.11

**Reply by dingo35** [reply]
I use same version.
Make sure you use the latest platformio.ini version from the repository!

**Comment by dingo35** [comment]
Why dont you all use the already compiled release: https://github.com/serkri/SmartEVSE-3/releases

**Comment by JohanEH22** [comment]
That's what I use now, and works. But I want to make some changes for my
particular situation so compiling is a necessity..

Op vr 1 dec. 2023 01:24 schreef dingo35 ***@***.***>:

**Reply by JohanEH22** [reply]
I did a complete fresh install on a new raspberry pi image:
installed platformio (latest)
downloaded master SEKRI git
platformio run
.
.
made a change in evse.cpp (doc["version"] = "johan" + String(VERSION);)
uploaded the firmware.bin via the browser /update 

no success. 

Something in the build works not ok, and it isnt the source (mstegen did also not work).

With:
esptool image_info --version 2 firmware.bin

and diff:

```
C:\Johan\_github_repos\Smartlastreleases>diff v173.txt sekri_joh.txt
2c2
< File size: 1169680 (bytes)
---
> File size: 1181024 (bytes)
8,9c8,9
< Entry point: 0x40083204
< Segments: 6
---
> Entry point: 0x400831dc
> Segments: 5
26,31c26,30
<       1  0x32178  0x3f400020  0x00000018  DROM
<       2  0x06d70  0x3ffbdb60  0x00032198  BYTE_ACCESSIBLE, DRAM
<       3  0x07100  0x40080000  0x00038f10  IRAM
<       4  0xcef60  0x400d0020  0x00040018  IROM
<       5  0x0e948  0x40087100  0x0010ef80  IRAM
<       6  0x00010  0x50000200  0x0011d8d0  RTC_DATA
---
>       1  0x32b80  0x3f400020  0x00000018  DROM
>       2  0x06664  0x3ffbdb60  0x00032ba0  BYTE_ACCESSIBLE, DRAM
>       3  0x06e04  0x40080000  0x0003920c  IRAM
>       4  0xd1608  0x400d0020  0x00040018  IROM
>       5  0x0ef00  0x40086e04  0x00111628  IRAM
35,36c34,35
< Checksum: 0x86 (valid)
< Validation hash: f44239444728b723066f04e01c133a27793fd9f7f0a993e0aa13928bcd1a70b8 (valid)
---
> Checksum: 0x8b (valid)
> Validation hash: c98077f9c955344726cbba126c4e0ae9253127b977db4b3b5681dabd09b7ad59 (valid)
41,44c40,43
< App version: esp-idf: v4.4.3 6407ecb3f8
< Compile time: Dec 20 2022      23:52:21
< ELF file SHA256: 4782bd80c20f9ddc5f2e9a1e6ad70da257f39306b7ca03a7df64415f87dfbaf0
< ESP-IDF: v4.4.3
---
> App version: esp-idf: v4.4.6 3572900934
> Compile time: Oct  4 2023      16:50:20
> ELF file SHA256: b4aa2cfc7e6445c8f1d058311089b4f8ad128a04171bdc88bb32a0d1657bd7e5
> ESP-IDF: v4.4.6-dirty
```

**Reply by JohanEH22** [reply]
```

C:\Users\Gebruiker\.platformio\penv\Scripts\python.exe esptool.py --chip esp32 image_info --version 2 C:\Johan\_github_repos\SmartEVSE_serkri\SmartEVSE-3\SmartEVSE-3\.pio\build\release\firmware.bin
esptool.py v4.5.1

ESP32 image header
==================
Image version: 1
Entry point: 0x400831dc
Segments: 5
Flash size: 4MB
Flash freq: 80m
Flash mode: DIO

ESP32 extended image header
===========================
WP pin: 0xee
Flash pins drive settings: clk_drv: 0x0, q_drv: 0x0, d_drv: 0x0, cs0_drv: 0x0, hd_drv: 0x0, wp_drv: 0x0
Chip ID: 0
Minimal chip revision: v0.0, (legacy min_rev = 0)
Maximal chip revision: v655.35

Segments information
====================
Segment   Length   Load addr   File offs  Memory types
-------  -------  ----------  ----------  ------------
      1  0x32b80  0x3f400020  0x00000018  DROM
      2  0x06664  0x3ffbdb60  0x00032ba0  BYTE_ACCESSIBLE, DRAM
      3  0x06e04  0x40080000  0x0003920c  IRAM
      4  0xd1608  0x400d0020  0x00040018  IROM
      5  0x0ef00  0x40086e04  0x00111628  IRAM

ESP32 image footer
==================
Checksum: 0xf1 (valid)
Validation hash: ac9055bfa56a76c92052741592ee77d3e96219e25fb7db80e398cdb564257982 (valid)

Application information
=======================
Project name: arduino-lib-builder
App version: esp-idf: v4.4.6 3572900934
Compile time: Oct  4 2023      16:50:20
ELF file SHA256: b5806b56c04c3a17835dcfc1572bccd3d2e1d147aed999c0597f7ed1af23f46b
ESP-IDF: v4.4.6-dirty
Secure version: 0


```

**Comment by negronoir** [comment]
I always have used your compiled versions. But I'd like to make some minor changes (e.g. minimum current for Tesla cars is 5A or change Webserver name for a more friendly one as I've two SmartEVSEs)

I can't find where's the problem. Because I've followed carefully the published 'building firmare' steps, and I can build img wo any problem both in Windows 10 & CentOS environments and upload to SmartEVSE. But none of them runs on target. After upload, only blank screen... (?)

**Comment by JohanEH22** [comment]
Ok, got the solution, arduinoespressif32 version 2.0.6 has idf 4.4.3, and it works like a charm (2.0.14 contains the 4.4.6-dirty). Maybe newer versions will also work, but I am ok with it for the time being :)

```
[env:release]
platform = espressif32
platform_packages =
   framework-arduinoespressif32 @ https://github.com/espressif/arduino-esp32#2.0.6
```

@negronoir could you confirm/mee eens?

**Comment by dingo35** [comment]
I can currently not test myself, but I can confirm that in my .platformio/packages, the version nr. of framework-arduinoespressif32 is 2.0.6  .

The change in the repository to 2.0.14 was reported to work succesfully by users, but again could not be tested by me personally.

So this might be the solution, Im awaiting confirmation of @negronoir .

**Reply by negronoir** [reply]
![image](https://github.com/serkri/SmartEVSE-3/assets/53239888/274f21ae-a940-469e-a700-4d04c4ee16ce)

It works! 

arduinoespressif32 version 2.0.6 has been the fix also for me.

Thanks both @dingo35 & @JohanEH22 for all the support.

**Reply by barteld** [reply]
Works for me as well!
This is on windows BTW.
Thanks all.

**Comment by dingo35** [comment]
Fixed in e629d8c20b4f

**Comment by JohanEH22** [comment]
I also tested it succesfully on Linux :)

Op wo 6 dec. 2023 15:09 schreef Prlwytzkofsky ***@***.***>:
